compiler/rustc_mir_transform/src/scalar_deref_prop.rs

```diff
@@ -0,0 +1,432 @@
+            _ => Value::Unknown,
+        };
+
+        targets[local] = rvalue;
+    }
+
+    debug!(?targets);
+
+    targets
+}
+
+/// Compute the set of locals that can be fully replaced.
+fn fully_replacable_locals(ssa: &SsaLocals, copies: &IndexVec<Local, Local>) -> BitSet<Local> {
+    let mut replacable = BitSet::new_empty(ssa.direct_uses.len());
+
+    // First pass: for each local, whether its uses can be fully replaced.
+    for (local, &num_direct_uses) in ssa.direct_uses.iter_enumerated() {
+        if num_direct_uses == 0 {
+            replacable.insert(local);
+        }
+    }
+
+    // Consolidate to have a local iff all its copies are.
+    for (local, &head) in copies.iter_enumerated() {
+        if !replacable.contains(local) {
+            replacable.remove(head);
+        }
+    }
+    for (local, &head) in copies.iter_enumerated() {
+        if !replacable.contains(head) {
+            replacable.remove(local);
+        }
+    }
+
+    // Verify that we correctly computed equivalence classes.
+    #[cfg(debug_assertions)]
+    for (local, &head) in copies.iter_enumerated() {
+        assert_eq!(replacable.contains(local), replacable.contains(head));
+    }
+
+    replacable
+}
+
+/// Utility to help performing subtitution of `*pattern` by `target`.
+struct Replacer<'tcx> {
+    tcx: TyCtxt<'tcx>,
+    targets: IndexVec<Local, Value<'tcx>>,
+    fully_replacable_locals: BitSet<Local>,
+    storage_to_remove: BitSet<Local>,
+}
+
+impl<'tcx> MutVisitor<'tcx> for Replacer<'tcx> {
+    fn tcx(&self) -> TyCtxt<'tcx> {
+        self.tcx
+    }
+
+    fn visit_local(&mut self, local: &mut Local, ctxt: PlaceContext, loc: Location) {
+        if !matches!(ctxt, PlaceContext::NonUse(_)) {
+            assert!(
+                !self.fully_replacable_locals.contains(*local),
+                "{local:?} should have been replaced at {loc:?} with context {ctxt:?}"
+            );
+        }
+    }
+
+    fn visit_place(&mut self, place: &mut Place<'tcx>, ctxt: PlaceContext, loc: Location) {
+        if let Value::Pointer(target) = self.targets[place.local]
+            && place.projection.first() == Some(&PlaceElem::Deref)
+        {
+            *place = target.project_deeper(&place.projection[1..], self.tcx);
+        }
+        self.super_place(place, ctxt, loc);
+    }
+
+    fn visit_assign(&mut self, place: &mut Place<'tcx>, rvalue: &mut Rvalue<'tcx>, loc: Location) {
+        if let Some(local) = place.as_local() && self.fully_replacable_locals.contains(local) {
+            // Ignore this statement, as it is dead code.
+            return;
+        }
+        self.super_assign(place, rvalue, loc);
+    }
+
+    fn visit_statement(&mut self, stmt: &mut Statement<'tcx>, loc: Location) {
+        if let StatementKind::StorageDead(l) = stmt.kind && self.storage_to_remove.contains(l) {
+            stmt.make_nop();
+        }
+        self.super_statement(stmt, loc);
+    }
```

Comment:
I think we should also remove the one and only assignment to `fully_replacable_locals`, even if this is resolved by other cleanup passes later.

That should allow you to remove the `visit_assign` above entirely